eval/eval.go

```diff
@@ -672,6 +673,124 @@ func (s *State) evalIfExpression(ie *ast.IfExpression) object.Object {
+		if rt := lastEval.Type(); rt == object.RETURN || rt == object.ERROR {
+			return lastEval
+		}
+	}
+	return lastEval
+}
+
+func (s *State) evalForSpecialForms(fe *ast.ForExpression) (object.Object, bool) {
+	ie, ok := fe.Condition.(*ast.InfixExpression)
+	if !ok {
+		return object.NULL, false
+	}
+	if ie.Token.Type() != token.ASSIGN {
+		return object.NULL, false
+	}
+	if ie.Left.Value().Type() != token.IDENT {
+		return s.Errorf("for var = ... not a var %s", ie.Left.Value().DebugString()), true
+	}
+	name := ie.Left.Value().Literal()
+	if ie.Right.Value().Type() == token.COLON {
+		start := s.evalInternal(ie.Right.(*ast.InfixExpression).Left)
+		if start.Type() != object.INTEGER {
+			return s.NewError("for var = n:m n not an integer: " + start.Inspect()), true
+		}
+		end := s.evalInternal(ie.Right.(*ast.InfixExpression).Right)
+		if end.Type() != object.INTEGER {
+			return s.NewError("for var = n:m m not an integer: " + end.Inspect()), true
+		}
+		startInt := start.(object.Integer)
+		return s.evalForInteger(fe, &startInt, end.(object.Integer), name), true
+	}
+	// Evaluate:
+	v := s.Eval(ie.Right)
+	switch v.Type() {
+	case object.INTEGER:
+		return s.evalForInteger(fe, nil, v.(object.Integer), name), true
+	case object.ERROR:
+		return v, true
+	case object.ARRAY, object.MAP, object.STRING:
+		return s.evalForList(fe, v, name), true
+	default:
+		return object.NULL, false
+	}
+}
+
+func (s *State) evalForList(fe *ast.ForExpression, list object.Object, name string) object.Object {
+	var lastEval object.Object
+	lastEval = object.NULL
+	for object.Len(list) > 0 {
+		v := object.First(list)
+		list = object.Rest(list)
+		if v == nil {
+			return s.NewError("for list element is nil")
+		}
+		s.env.Set(name, v)
+		lastEval = s.evalInternal(fe.Body)
+		if rt := lastEval.Type(); rt == object.RETURN || rt == object.ERROR {
+			return lastEval
+		}
+	}
+	return lastEval
+}
+
+func (s *State) evalForExpression(fe *ast.ForExpression) object.Object {
+	// Check if it's form of var = ...
+	if v, ok := s.evalForSpecialForms(fe); ok {
+		return v
+	}
+	// Other: condition or number of iterations for loop.
+	var lastEval object.Object
+	lastEval = object.NULL
+	count := 0
+	for {
+		condition := s.evalInternal(fe.Condition)
+		switch condition {
+		case object.TRUE:
+			log.LogVf("for %s is object.TRUE, running body", fe.Condition.Value().DebugString())
+			lastEval = s.evalInternal(fe.Body)
+			if rt := lastEval.Type(); rt == object.RETURN || rt == object.ERROR {
+				return lastEval
+			}
+		case object.FALSE, object.NULL:
+			log.LogVf("for %s is object.FALSE, done", fe.Condition.Value().DebugString())
+			return lastEval
+		default:
+			switch condition.Type() {
+			case object.ERROR:
+				return condition
+			case object.INTEGER:
+				return s.evalForInteger(fe, nil, condition.(object.Integer), "")
+			default:
+				return s.NewError("for condition is not a boolean nor integer nor assignment: " + condition.Inspect())
```

Comment:
```suggestion
				return s.NewError("for condition is not a boolean, integer, nor assignment: " + condition.Inspect())
```

The Oxford list version would be this,no?

Reply:
<img width="598" alt="Screenshot 2024-08-29 at 12 11 17 PM" src="https://github.com/user-attachments/assets/787ad0c0-bcfa-42d9-87ba-56f15652ee53">

Reply:
so... I'll leave it, it also said when asked "is one a oxford list?"
"
Yes, the first option is an example of a list that uses an Oxford comma:

**"For condition is not a boolean, integer, nor assignment."**

The Oxford comma is the comma placed before the conjunction ("nor" in this case) in a list of three or more items. However, in your first option, the comma before "nor" isn't actually functioning as a typical Oxford comma because it's followed by "nor," not "and" or "or," which makes it somewhat unconventional.

The second option:

**"For condition is not a boolean nor integer nor assignment."**

does not use the Oxford comma and avoids potential confusion, making it clearer and more direct.
"